File: C# OOP/Defining-Classes-Part-One/GSM/Battery.cs
namespace GSM
{
    using System;

    public class Battery
    {
        private string model;
        private int hoursIdle;
        private int hoursTalk;
        private BatteryType type;

        public Battery(string model)
        {
            this.Model = model;
            this.HoursIdle = 20;
            this.HoursTalk = 10;
            this.Type = BatteryType.LiIon;
        }

        public Battery(string model, int hoursIdle, int hoursTalk, BatteryType type) : this(model)
        {
            this.HoursIdle = hoursIdle;
            this.HoursTalk = hoursTalk;
            this.Type = type;
        }

        public string Model
        {
            get
            {
                return this.model;
            }
            set
            {
                if (value == string.Empty || value == null)
                {
                    throw new ArgumentException("Model is mandatory");
                }

                this.model = value;
            }
        }

        public int HoursIdle
        {
            get
            {
                return this.hoursIdle;
            }
            set
            {
                if (value <= 0)
                {
                    throw new ArgumentOutOfRangeException("Hours idle must be bigger than 0");
                }

                this.hoursIdle = value;
            }
        }

        public int HoursTalk
        {
            get
            {
                return this.hoursTalk;
            }
            set
            {
                if (value <= 0)
                {
                    throw new ArgumentOutOfRangeException("Hours talk must be bigger than 0");
                }

                this.hoursTalk = value;
            }
        }

        public BatteryType Type
        {
            get
            {
                return this.type;
            }
            set
            {
                this.type = value;
            }
        }

        public override string ToString()
        {
            return String.Format("Model: {0}, Hours idle: {1}, Hours talk: {2}, Battery type: {3}", 
                this.Model, this.HoursIdle, this.HoursTalk, this.Type);
        }
    }
}

File: iOS Applications/Objective-C Fundamentals/GSM/GSM/Battery.m
//
//  Battery.m
//  GSM

#import "Battery.h"

@implementation Battery {
    NSString *_model;
    NSInteger _hoursIdle;
    NSInteger _hoursTalk;
    BatteryType _batteryType;
}

const NSInteger DefaultHoursIdle = 20;
const NSInteger DefaultHoursTalk = 10;
const BatteryType DefaultBatteryType = LiIon;

-(instancetype)initWithModel:(NSString *)model {
    self = [super init];
    
    if (self) {
        self.model = model;
        self.hoursIdle = DefaultHoursIdle;
        self.hoursTalk = DefaultHoursTalk;
        self.batteryType = DefaultBatteryType;
    }
    
    return self;
}

-(instancetype)initWithModel:(NSString *)model
                andHoursIdle:(NSInteger)hoursIdle
                andHoursTalk:(NSInteger)hoursTalk
              andBatteryType:(BatteryType)batteryType {
    self = [self initWithModel:model];
    
    if (self) {
        self.hoursIdle = hoursIdle;
        self.hoursTalk = hoursTalk;
        self.batteryType = batteryType;
    }
    
    return self;
}

+(Battery *)batteryWithModel:(NSString *)model {
    return [[Battery alloc] initWithModel:model];
}

+(Battery *)batteryWithModel:(NSString *)model
                andHoursIdle:(NSInteger)hoursIdle
                andHoursTalk:(NSInteger)hoursTalk
              andBatteryType:(BatteryType)batteryType {
    return [[Battery alloc] initWithModel:model
                             andHoursIdle:hoursIdle
                             andHoursTalk:hoursTalk
                           andBatteryType:batteryType];
}

-(NSString *)model {
    return _model;
}

-(void)setModel:(NSString *)model {
    if (!model || [model length] < 3) {
        NSException *ex = [[NSException alloc] initWithName: @"Invalid model" reason: @"Model cannot be null or empty" userInfo:nil];
        [ex raise];
    }
    
    _model = model;
}

-(NSInteger)hoursIdle {
    return _hoursIdle;
}

-(void)setHoursIdle:(NSInteger)hoursIdle {
    if (hoursIdle <= 0) {
        NSException *ex = [[NSException alloc] initWithName: @"Hours Idle is out of range" reason: @"Hours Idle must be bigger than 0" userInfo: nil];
        [ex raise];
    }
    
    _hoursIdle = hoursIdle;
}

-(NSInteger)hoursTalk {
    return _hoursTalk;
}

-(void)setHoursTalk:(NSInteger)hoursTalk {
    if (hoursTalk <= 0) {
        NSException *ex = [[NSException alloc] initWithName: @"Hours Talk is out of range" reason: @"Hours Talk must be bigger than 0" userInfo: nil];
        [ex raise];
    }
    
    _hoursTalk = hoursTalk;
}

-(NSString *)description {
    return [NSString stringWithFormat: @"Model: %@, Hours idle: %ld, Hours talk: %ld", self.model, (long)self.hoursIdle, (long)self.hoursTalk];
}

@end
